drivers/vdpa/virtio_mi/lm.c

```diff
@@ -41,6 +41,213 @@ TAILQ_HEAD(virtio_vdpa_mi_privs, virtio_vdpa_pf_priv) virtio_mi_priv_list =
+				numa_node);
+	if (vq == NULL) {
+		DRV_LOG(ERR, "can not allocate vq %u", queue_idx);
+		return -ENOMEM;
+	}
+	hw->vqs[queue_idx] = vq;
+
+	vq->hw = hw;
+	vq->vq_queue_index = queue_idx;
+	vq->vq_nentries = vq_size;
+
+	/*
+	 * Reserve a memzone for vring elements
+	 */
+	size = vring_size(hw, vq_size, VIRTIO_VRING_ALIGN);
+	vq->vq_ring_size = RTE_ALIGN_CEIL(size, VIRTIO_VRING_ALIGN);
+
+	mz = rte_memzone_reserve_aligned(vq_name, vq->vq_ring_size,
+			numa_node, RTE_MEMZONE_IOVA_CONTIG,
+			VIRTIO_VRING_ALIGN);
+	if (mz == NULL) {
+		if (rte_errno == EEXIST)
+			mz = rte_memzone_lookup(vq_name);
+		if (mz == NULL) {
+			ret = -ENOMEM;
+			goto err_ret;
+		}
+	}
+
+	memset(mz->addr, 0, mz->len);
+
+	vq->vq_ring_mem = mz->iova;
+	vq->vq_ring_virt_mem = mz->addr;
+
+	virtio_vdpa_init_vring(vq);
+
+	if (queue_idx == (priv->vpdev->common_cfg->num_queues - 1)) {
+		avq = &vq->aq;
+		avq->mz = mz;
+
+		/* Allocate a page for admin vq command, data and status */
+		sz_hdr_mz = rte_mem_page_size();
+
+		if (sz_hdr_mz) {
+			snprintf(vq_hdr_name, sizeof(vq_hdr_name), "vdev%d_vq%u_hdr",
+					vpdev->vfio_dev_fd, queue_idx);
+			hdr_mz = rte_memzone_reserve_aligned(vq_hdr_name, sz_hdr_mz,
+					numa_node, RTE_MEMZONE_IOVA_CONTIG,
+					RTE_CACHE_LINE_SIZE);
+			if (hdr_mz == NULL) {
+				if (rte_errno == EEXIST)
+					hdr_mz = rte_memzone_lookup(vq_hdr_name);
+				if (hdr_mz == NULL) {
+					ret = -ENOMEM;
+					goto free_mz;
+				}
+			}
+			avq->virtio_admin_hdr_mz = hdr_mz;
+			avq->virtio_admin_hdr_mem = hdr_mz->iova;
+			memset(avq->virtio_admin_hdr_mz->addr, 0, rte_mem_page_size());
+		} else {
+			DRV_LOG(ERR, "rte mem page size is zero");
+		}
+
+		hw->avq = avq;
+	}
+
+	vr_info.size  = vq_size;
+	vr_info.desc  = (uint64_t)(uintptr_t)vq->vq_split.ring.desc;
+	vr_info.avail = (uint64_t)(uintptr_t)vq->vq_split.ring.avail;
+	vr_info.used  = (uint64_t)(uintptr_t)vq->vq_split.ring.used;
+	ret = virtio_pci_dev_queue_set(vpdev, queue_idx, &vr_info);
+	if (ret) {
+		DRV_LOG(ERR, "setup_queue %u failed", queue_idx);
+		ret = -EINVAL;
+		goto clean_vq;
+	}
+
+	return 0;
+
+clean_vq:
+	hw->avq = NULL;
+	rte_memzone_free(hdr_mz);
+free_mz:
+	rte_memzone_free(mz);
+err_ret:
+	return ret;
+}
+
+static void
+virtio_vdpa_admin_queue_free(struct virtio_vdpa_pf_priv *priv)
+{
+	uint16_t nr_vq = priv->hw_nr_virtqs;
+	struct virtio_hw *hw = &priv->vpdev->hw;
+	struct virtqueue *vq;
+	uint16_t i;
+
+	if (hw->vqs == NULL)
+		return;
+
+	if (hw->avq) {
+		virtio_vdpa_destroy_aq_ctl(hw->avq);
+		hw->avq = NULL;
+	}
+
+	for (i = 0; i < nr_vq; i++) {
+		vq = hw->vqs[i];
+		if (vq) {
+			rte_free(vq);
+			hw->vqs[i] = NULL;
+		}
+	}
+	rte_free(hw->vqs);
+	hw->vqs = NULL;
+}
+
+static int
+virtio_vdpa_admin_queue_alloc(struct virtio_vdpa_pf_priv *priv)
+{
+	struct virtio_hw *hw = &priv->vpdev->hw;
+	uint16_t i, queue_idx;
+	int ret;
+
+	hw->max_queue_pairs = virtio_pci_dev_nr_vq_get(priv->vpdev) / 2;
+	priv->hw_nr_virtqs = 1;
```

Comment:
use callback function to net and blk. get admin queue indx.
then init admin queue

Reply:
Done